File: internal/sync/devbox.go
package sync

import (
	"context"
	"strings"

	"github.com/iskry/cask/internal/config"
	"github.com/iskry/cask/internal/executor"
)

// DevboxSyncManager synchronizes distrobox instances between host and config.
type DevboxSyncManager struct {
	Config *config.DevboxConfig
}

func (m *DevboxSyncManager) ResourceType() string { return "devboxes" }

func (m *DevboxSyncManager) GetHostResources(ctx context.Context, exec executor.Executor) ([]Resource, error) {
	r, err := exec.Execute(ctx, []string{"distrobox", "list", "--no-color"})
	if err != nil {
		return nil, err
	}
	if !r.Success {
		return nil, nil
	}
	var resources []Resource
	lines := strings.Split(r.Stdout, "\n")
	for _, line := range lines[1:] { // Skip header
		parts := strings.Split(line, "|")
		if len(parts) >= 3 {
			resources = append(resources, Resource{
				"id":     strings.TrimSpace(parts[0]),
				"name":   strings.TrimSpace(parts[1]),
				"status": strings.TrimSpace(parts[2]),
			})
		}
	}
	return resources, nil
}

func (m *DevboxSyncManager) GetConfigResources() []Resource {
	if m.Config == nil {
		return nil
	}
	var resources []Resource
	for _, inst := range m.Config.Instances {
		resources = append(resources, Resource{
			"name":  inst.Name,
			"image": inst.Image,
		})
	}
	return resources
}

func (m *DevboxSyncManager) ResourceID(r Resource) string {
	name, _ := r["name"].(string)
	return name
}

func (m *DevboxSyncManager) Apply(ctx context.Context, exec executor.Executor, r Resource) bool {
	name, _ := r["name"].(string)
	image, _ := r["image"].(string)
	result, err := exec.Execute(ctx, []string{
		"distrobox", "create", "--name", name, "--image", image, "--yes",
	})
	return err == nil && result.Success
}

func (m *DevboxSyncManager) Remove(ctx context.Context, exec executor.Executor, id string) bool {
	r, err := exec.Execute(ctx, []string{"distrobox", "rm", "--force", id})
	return err == nil && r.Success
}

func (m *DevboxSyncManager) NeedsUpdate(host, cfg Resource) bool {
	hostImage, _ := host["image"].(string)
	cfgImage, _ := cfg["image"].(string)
	return hostImage != cfgImage
}

File: src/cask/sync/devbox.py
"""Distrobox sync implementation."""
from __future__ import annotations

from dataclasses import dataclass, field

from cask.config.models import DevboxConfig, DevboxInstanceConfig
from cask.executor.protocol import Executor
from cask.managers.distrobox import DistroboxManager
from cask.result import Result


@dataclass
class DevboxResource:
    name: str
    config: DevboxInstanceConfig | None = None
    image: str = ""
    packages: list[str] = field(default_factory=list)


class DevboxSync:
    def __init__(self) -> None:
        self._mgr = DistroboxManager()

    async def get_host_resources(self, exec: Executor) -> list[DevboxResource]:
        instances = await self._mgr.list_instances(exec)
        return [DevboxResource(name=n, image=img) for n, img in instances.items()]

    def get_config_resources(self, config: DevboxConfig) -> list[DevboxResource]:
        return [
            DevboxResource(name=n, config=c, image=c.image, packages=list(c.packages))
            for n, c in config.instances.items()
        ]

    async def apply(self, resource: DevboxResource, exec: Executor) -> Result:
        if resource.config:
            return await self._mgr.create(resource.name, resource.config, exec)
        return Result(ok=False, message=f"No config for {resource.name}", actions=[])

    async def remove(self, resource_id: str, exec: Executor) -> Result:
        return await self._mgr.remove(resource_id, exec)

    def needs_update(self, host: DevboxResource, config: DevboxResource) -> bool:
        return host.image != config.image

    def resource_id(self, resource: DevboxResource) -> str:
        return resource.name
